### PR TITLE
FLEX-5445: Support for several messages of the same type for one mock…

### PR DESCRIPTION
**Pull request by joostknapen**
…ed device. Support for several mocked devices. Initial commit to see which cucumber test fail during build

### CODE REVIEW

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java [hunk 1]
```diff
@@ -149,15 +146,18 @@ public MockOslpServer(final CoreDeviceConfiguration configuration, final int osl
         this.responseDelayRandomRange = reponseDelayRandomRange;
     }
 
-    public Integer getSequenceNumber() {
-        return this.channelHandler.getSequenceNumber();
+    public Integer getSequenceNumber(final String deviceUID) {
```
Comment (bvdzwet) [inline integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java:149]: Proper camel-casing would be `deviceUid`. (Also elsewhere in other added methods.)

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java [hunk 2]
```diff
@@ -180,23 +180,23 @@ public void stop() {
...
         int count = 0;
-        while (!this.receivedRequests.containsKey(requestType)) {
+        while (!deviceState.hasReceivedRequests(messageType)) {
             try {
                 count++;
-                LOGGER.info("Sleeping 1s " + count + " - waiting for request of message type: " + requestType.name());
+                LOGGER.info("Sleeping 1s {} - Device {} is waiting for a request of message type: {}", count, deviceUID, messageType.name());
                 Thread.sleep(1000);
             } catch (final InterruptedException e) {
                 Assertions.fail("Polling for response interrupted");
```
Comment (bvdzwet) [inline integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java:202]: Not from this pull request, but the failure message seems deceptive. The waiting here looks like it is for a received request, not for a response. (The same applies just a few lines below with the next `fail`.)

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/resources/features/publiclighting/osgp-adapter-ws-core/ConfigurationManagement/SetConfiguration.feature [hunk 3]
```diff
@@ -154,7 +152,7 @@ Feature: CoreConfigurationManagement SetConfiguration
       | DeviceIdentification | TEST1024000000001 |
       | Protocol             | <Protocol>        |
     And the device returns a set configuration status over "<Protocol>"
-      | Status            | FAILURE   |
+      | Status            | FAILURE  |
```
Comment (bvdzwet) [inline integration-tests/cucumber-tests-platform-publiclighting/src/test/resources/features/publiclighting/osgp-adapter-ws-core/ConfigurationManagement/SetConfiguration.feature:155]: Why break the alignment of the "closing bars"?

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/resources/features/publiclighting/osgp-adapter-ws-core/ConfigurationManagement/SetConfiguration.feature [hunk 4]
```diff
@@ -185,13 +183,13 @@ Feature: CoreConfigurationManagement SetConfiguration
...
 
-  @OslpMockServer
+  @OslpMockServer @SetConfigurationRejected
   Scenario Outline: Rejected set configuration of a device
     Given an ssld oslp device
       | DeviceIdentification | TEST1024000000001 |
       | Protocol             | <Protocol>        |
     And the device returns a set configuration status over "<Protocol>"
-      | Status            | REJECTED  |
+      | Status            | REJECTED   |
```
Comment (bvdzwet) [inline integration-tests/cucumber-tests-platform-publiclighting/src/test/resources/features/publiclighting/osgp-adapter-ws-core/ConfigurationManagement/SetConfiguration.feature:192]: Why break the alignment of the "closing bars"?

**Review by bvdzwet — CHANGES_REQUESTED** [review]

File: integration-tests/cucumber-tests-platform-publiclighting/pom.xml [hunk 5]
```diff
@@ -113,6 +113,14 @@
       <scope>provided</scope>
     </dependency>
 
+    <!-- Lombok -->
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/pom.xml:116]: I agree with you on that. I saw comments like this on multiple places in pom's and basically I just copied it. I will remove it, dependent if that's in line with the rest of this particular pom

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java [hunk 6]
```diff
@@ -106,12 +106,27 @@
      *            device.
      */
     @Then("^a get configuration \"([^\"]*)\" message is sent to device \"([^\"]*)\"$")
-    public void aGetConfigurationOSLPMessageIsSentToDevice(final String protocol, final String deviceIdentification) {
-        final Message message = this.oslpMockServer.waitForRequest(MessageType.GET_CONFIGURATION);
+    public void aGetConfigurationOSLPMessageIsSentToDevice(final String protocol, final String deviceIdentification)
+            throws DeviceSimulatorException {
+        this.aGetConfigurationOSLPMessageIsSentToSpecificDevice(protocol, deviceIdentification, this.getDeviceUID(new HashMap<>()));
+    }
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:112]: I will refactor all to getDeviceUid

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java [hunk 7]
```diff
@@ -106,12 +106,27 @@
...
+
+    /**
+     * Verify that a get configuration OSLP message is sent to the specific device.
+     *
+     * @param deviceIdentification
+     *            The device identification expected in the message to the
+     *            device.
+     */
+    @Then("^a get configuration \"([^\"]*)\" message is sent to device \"([^\"]*)\" with deviceUid \"([^\"]*)\"$")
+    public void aGetConfigurationOSLPMessageIsSentToSpecificDevice(final String protocol, final String deviceIdentification, final String deviceUID)
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:122]: As how things are build in our cucumber test set your statement is not true. (I am not saying that you are wrong, but that's not how the cucumber test are set up)

The default device_uid as set in PlatformDefaults.java = dGVzdDEyMzQ1Njc4. That decodes to test12345678
The default device_identifcation as set in PlatformDefaults.java = TD01

So in the tests device_uid and device_identifacation are not coupled in any way. If we do want to couple them, that will likely requires changes in many tests. I did not look into that. I also wa not aware of the (pretented) relatation between device_uid and device_identification. 

For now, I will suggest to keep the fields unrelated. Maybe it should be another issue on the back log.
What do you think?

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java [hunk 8]
```diff
@@ -106,12 +106,27 @@
...
+            throws DeviceSimulatorException {
+        this.aGetConfigurationOSLPMessageIsSentToSpecificDevice(protocol, deviceIdentification, this.getDeviceUID(new HashMap<>()));
+    }
+
+    /**
+     * Verify that a get configuration OSLP message is sent to the specific device.
+     *
+     * @param deviceIdentification
+     *            The device identification expected in the message to the
+     *            device.
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:119]: No, I guess this is old code. There are, unfortunately, many flaws in the Java documentation. I can have a look at this case.

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java [hunk 9]
```diff
@@ -106,12 +106,27 @@
      *            device.
      */
     @Then("^a get configuration \"([^\"]*)\" message is sent to device \"([^\"]*)\"$")
-    public void aGetConfigurationOSLPMessageIsSentToDevice(final String protocol, final String deviceIdentification) {
-        final Message message = this.oslpMockServer.waitForRequest(MessageType.GET_CONFIGURATION);
+    public void aGetConfigurationOSLPMessageIsSentToDevice(final String protocol, final String deviceIdentification)
+            throws DeviceSimulatorException {
+        this.aGetConfigurationOSLPMessageIsSentToSpecificDevice(protocol, deviceIdentification, this.getDeviceUID(new HashMap<>()));
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:111]: Agree with you on that. I had this discussion before with Kevin. I already proposed to remove all these unused parameters. Removing all these protocol parameters should in my opionion be a new story on the back log.

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java (hunk 1 above)
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java:149]: I will refactor this

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java (hunk 2 above)
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpServer.java:202]: true :-) will change this

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (bvdzwet) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:122]: Agreed it would probably have too much impact to do something about the device identification vs UID split to make the tests simpler (if the assumptions hold up we can regard the relation between the two as fixed).

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:112]: I am not sure how much was already there --> most of it was already there ;-)

File: integration-tests/cucumber-tests-platform-publiclighting/pom.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/pom.xml:116]: I removed the comments line. Although many, but not all, dependencies in the pom still have comments

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:119]: I just checked. All methods that have 'protocol' as a parameter, are missing this comment. Ugly, but not really related to this issue. Maybe fix it if we decide to remove the protocol parameter. (Then is it automatically fixed ;-) )

[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:119]: Checked... This is a particularly bad case. The method has 3 parameters of which 2 are not used. (Changed the doc)

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java [hunk 10]
```diff
@@ -886,55 +1006,89 @@ public void theDeviceReturnsASetLightScheduleResponseOverOSLP(final String resul
...
     }
 
+
     /**
      * Setup method to stop a device which should be returned by the mock.
      */
     @Given("^the device returns a stop device response \"([^\"]*)\" over \"([^\"]*)\"$")
     public void theDeviceReturnsAStopDeviceResponseOverOSLP(final String result, final String protocol) {
         // TODO: Check if ByteString.EMPTY must be something else
-        this.oslpMockServer.mockStopDeviceResponse(ByteString.EMPTY, Enum.valueOf(Status.class, result));
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:937]: Removed the TODO

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java [hunk 11]
```diff
@@ -1250,4 +1423,17 @@ private static String convertIpAddress(final ByteString byteString) {
...
     }
+
+    private Map<String, String> setDeviceUid(final Map<String, String> settings, final String deviceUid) {
+        settings.put(PlatformPubliclightingKeys.KEY_DEVICE_UID, deviceUid);
+        return settings;
+    }
+
+    private String getDeviceUID(Map<String, String> settings) {
+        if (settings == null) {
+            settings = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:1434]: Maybe nicer to just return the default DEVICE_UID instead --> That's what it does. I use this method as helper method to get rid of this code that is used in may places (which this method encapsulates) 
       
return getString(settings, PlatformPubliclightingKeys.KEY_DEVICE_UID,
                PlatformPubliclightingDefaults.DEVICE_UID);

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2020 Smart Society Services B.V.
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:2]: Oops

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+
+public class DeviceState {
+
+    @Getter
+    private final String deviceUID;
+    @Getter @Setter
+    private Integer sequenceNumber = 0;
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:28]: Good point, but in this case, I don't think it really is an issue. There is only one class from where the sequencenumber is retrieved or changed, and that is from MockOslpChannelHandler. There is only one instance of that class, and that's not shared over threads.

The fact that ConcurrentMaps are used, is because there is a mock server thread, and a 'cucumber' thread. The cucumber thread adds mocks to the map, and the mock server removes them.

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+        this.sequenceNumber = next;
+    }
+
+    private ConcurrentLinkedQueue<Oslp.Message> getQueue(final Map<MessageType, ConcurrentLinkedQueue<Oslp.Message>> messageMap, final MessageType messageType) {
+        messageMap.computeIfAbsent(messageType, k -> new ConcurrentLinkedQueue<>());
+        return messageMap.get(messageType);
+    }
+
+    private Oslp.Message getFromMap(final Map<MessageType, ConcurrentLinkedQueue<Oslp.Message>> messageMap, final MessageType messageType)
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:95]: Hmm, indeed, the name is unclear. I changed it to pollMessage. Since that's what is does. It polls the queue (picks up first item and removes it from the queue)

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    public Oslp.Message getReceivedRequest(final MessageType messageType) throws DeviceSimulatorException {
+        return this.getFromMap(this.receivedRequestsMap, messageType);
+    }
+
+    public void addMockedResponse(final MessageType messageType, final Oslp.Message message) {
+        this.getQueue(this.mockedResponsesMap, messageType).add(message);
+    }
+
+    public void addReceivedRequest(final MessageType messageType, final Oslp.Message message) {
+        this.getQueue(this.receivedRequestsMap, messageType).add(message);
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:52]: I go with you for getNextMessage, but not with addNextMessage. Because when you add a message, getNextMessage can return a different message. I changed the getters to poll<xx>

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (bvdzwet) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:1434]: Regarding just returning the default device UID, that is _not_ what it does. More precisely, that is not what the line a commented on does, it _is_ what the method does, but that wasn't my point.
What it does, is producing some return value by assigning to a parameter (bad practice), which causes a side effect later-on.
The method is rather short, so it is not that hard to comprehend, and you can leave it as is if you like it better this way.

[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/glue/steps/mocks/OslpDeviceSteps.java:1434]: Oh, agreed, thanks for pointing out. It can potentially hide an unwanted edge case (cq bug) where settings == null
I'll remove the null check since it is/should not be needed anyway.

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (bvdzwet) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:52]: Ha, I had considered something with poll as a suggestion but rejected it because I felt it was too much about a non-essential implementation aspect. I am OK if you stick with the poll-based naming though, as I think it isn't a very clear cut case which is better.

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DevicesContext.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    private final ConcurrentMap<String, DeviceState> mockedDevicesMap = new ConcurrentHashMap<>();
+
+    public DeviceState getDeviceState(final String deviceIdentification) {
+        DeviceState deviceState = this.mockedDevicesMap.get(deviceIdentification);
+        if (deviceState == null) {
+            deviceState = new DeviceState(deviceIdentification);
+            this.mockedDevicesMap.put(deviceIdentification, deviceState);
+        }
+
+        return deviceState;
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DevicesContext.java:24]: I guess so ;-) thanks

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DevicesContext.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+package org.opensmartgridplatform.cucumber.platform.publiclighting.mocks.oslpdevice;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class DevicesContext {
+
+    private final ConcurrentMap<String, DeviceState> mockedDevicesMap = new ConcurrentHashMap<>();
+
+    public DeviceState getDeviceState(final String deviceIdentification) {
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DevicesContext.java:17]: Thanks, changed it

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpChannelHandler.java [hunk 18]
```diff
@@ -228,35 +223,41 @@ public void channelRead0(final ChannelHandlerContext ctx, final OslpEnvelope mes
...
                     // Build the OslpEnvelope.
                     final OslpEnvelope.Builder responseBuilder = new OslpEnvelope.Builder()
                             .withSignature(this.oslpSignature)
                             .withProvider(this.oslpSignatureProvider)
                             .withPrimaryKey(this.privateKey)
-                            .withDeviceId(deviceId);
+                            .withDeviceId(message.getDeviceId());
 
                     // Pass the incremented sequence number to the
                     // handleRequest() function for checking.
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpChannelHandler.java:243]: Hmm, I had a look. What the code actually does it at first glance not really clear. The handleRequest(message) does increment the sequence number, but there are no checks in relation to the incoming message.

[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpChannelHandler.java:243]: I removed the comments

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpChannelHandler.java [hunk 19]
```diff
@@ -228,35 +223,41 @@ public void channelRead0(final ChannelHandlerContext ctx, final OslpEnvelope mes
...
-                            convertByteArrayToInteger(response.getSequenceNumber()));
+                    LOGGER.debug("Device {} is sending an OSLP response with sequence number {}",
+                           this.getDeviceUID(response), convertByteArrayToInteger(response.getSequenceNumber()));
 
                     // wait for the response to actually be written. This
                     // improves stability of the tests
                     final ChannelFuture future = ctx.channel().writeAndFlush(response);
                     future.await();
 
                     LOGGER.debug("Send OSLP Response: {}", response.getPayloadMessage().toString().split(" ")[0]);
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/MockOslpChannelHandler.java:258]: Indeed

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public boolean hasReceivedRequests(final MessageType messageType) {
+        return !this.getQueue(this.receivedRequestsMap, messageType).isEmpty();
+    }
+
+    public void incrementSequenceNumber() {
+        int numberToAddToSequenceNumberValue = 1;
+
+        if (ScenarioContext.current().get(PlatformKeys.NUMBER_TO_ADD_TO_SEQUENCE_NUMBER) != null) {
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:66]: To be honest, I just copied this whole method from (I thought the handler), and didn't change anything. It wasn't really clear to me what all exactly it should to, so I left it unaffected. I'll have a look if I can refactor the method.

[the remark this answers is not in the document]
Comment (bvdzwet) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:66]: OK, just see what quick improvements you could do, if any, With some code being moved around (which is fine) I did not see immediately what was new, and what was "inherited" from whoever worked on it before.

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        int next = this.sequenceNumber + numberToAddToSequenceNumberValue;
+        if (next > SEQUENCE_NUMBER_MAXIMUM) {
+            final int sequenceNumberMaximumCross = next - SEQUENCE_NUMBER_MAXIMUM;
+            if (sequenceNumberMaximumCross >= 1) {
+                next = sequenceNumberMaximumCross - 1;
+            }
+        } else if (next < 0) {
+            final int sequenceNumberMaximumCross = next * -1;
+            if (sequenceNumberMaximumCross >= 1) {
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:82]: agreed

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+                    .get(PlatformKeys.NUMBER_TO_ADD_TO_SEQUENCE_NUMBER)
+                    .toString();
+            if (!numberToAddAsNextSequenceNumber.isEmpty()) {
+                numberToAddToSequenceNumberValue = Integer.parseInt(numberToAddAsNextSequenceNumber);
+            }
+        }
+        int next = this.sequenceNumber + numberToAddToSequenceNumberValue;
+        if (next > SEQUENCE_NUMBER_MAXIMUM) {
+            final int sequenceNumberMaximumCross = next - SEQUENCE_NUMBER_MAXIMUM;
+            if (sequenceNumberMaximumCross >= 1) {
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:77]: agreed

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+            final String numberToAddAsNextSequenceNumber = ScenarioContext.current()
+                    .get(PlatformKeys.NUMBER_TO_ADD_TO_SEQUENCE_NUMBER)
+                    .toString();
+            if (!numberToAddAsNextSequenceNumber.isEmpty()) {
+                numberToAddToSequenceNumberValue = Integer.parseInt(numberToAddAsNextSequenceNumber);
+            }
+        }
+        int next = this.sequenceNumber + numberToAddToSequenceNumberValue;
+        if (next > SEQUENCE_NUMBER_MAXIMUM) {
+            final int sequenceNumberMaximumCross = next - SEQUENCE_NUMBER_MAXIMUM;
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:76]: Looks like a check for an overflow

File: integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+            if (sequenceNumberMaximumCross >= 1) {
+                next = sequenceNumberMaximumCross - 1;
+            }
+        } else if (next < 0) {
+            final int sequenceNumberMaximumCross = next * -1;
+            if (sequenceNumberMaximumCross >= 1) {
+                next = SEQUENCE_NUMBER_MAXIMUM - sequenceNumberMaximumCross + 1;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:85]: In case that is the intended behaviour :-) After some simple refactoring I came to this if/else simplifcation:
        int next = this.sequenceNumber + increment;
        if (next > SEQUENCE_NUMBER_MAXIMUM) {
            next -= SEQUENCE_NUMBER_MAXIMUM - 1;
        } else if (next < 0) {
            next = SEQUENCE_NUMBER_MAXIMUM - next * -1 + 1;
        }
which is kinda weird... why an else? how could next become < 0? I guess only if you set a negative increment. But why would you do that? Is that possible?

[the remark this answers is not in the document]
Comment (joostknapen) [reply integration-tests/cucumber-tests-platform-publiclighting/src/test/java/org/opensmartgridplatform/cucumber/platform/publiclighting/mocks/oslpdevice/DeviceState.java:85]: ProtocolSequenceNumber.feature covers tests where a negative increment is indeed possible.... The test does not cover edge cases. But for this reason I added the else again.

**Review by bvdzwet — APPROVED** [review]
